### DISCUSSION TITLE
strategies to improve build times in Android

### DISCUSSION THREAD

**Discussion by manask88** [MobileNativeFoundation/discussions#56] [General]
I wonder what have others done in order to improve their build times. At Pinterest, we are currently working towards modularization, which is probably the number one lever in order to improve build times. However, this a long process, so at the same time, we've also been looking for quick wins to improve build time.

This is our current setup:

gradle 6.6, kotlin 1.4.10 and ag 4.1.2, we'll soon update to gradle 6.8, which will give us configuration cache and the ability to ignore empty folders

almost 900K lines of code, where 2/3 is kotlin. We have about 20 modules, but a really big one where most of the changes happen

Some of the things we've done recently on different fronts:

Improve Remote cache hits:
- Make sure to remove as much nondeterministic code or code that changes when builds runs on CI
- Make sure to remove empty folders

Improve local cache hits
- enable kapt build cache

Improve incremental builds:
- enable incremental kapt builds, making sure that all your incremental kapt processors support incremental builds
- create kotlin files in java only projects due to this bug https://youtrack.jetbrains.com/issue/KT-34862

However, we still see issues on a couple of fronts:

- remote cache is very sensitive. Minimal changes / unrelated changes can affect kapt generated class method orders, which then causes lots of cache misses. Some offending changes I've seen for this:
       * using a different java version due to using a preview version of AS (Java 11 instead of java 8)
       * having empty folders

- we are not able to update to kotlin 1.4.31 given this bug:  https://youtrack.jetbrains.com/issue/KT-43686

- incremental compilation a lot of times fails and instead, the compiler goes for full compilation. I have been able to reproduce this consistently under the following scenarios. 
      * when cache is enabled and there is a non abi change
      * when cache is enabled and there is an abi change
      * when cache is on or off and there a resource id addition/removal

**Comment by roscrazy** [comment]
Hi, I am working on a project with more than 2.5 million LOC, ~1k modules. From your question, I would like to share a bit on what we've been doing to improve the build time for our project during last year.

- Build a critical path dashboard, and focus on optimizing the critical path by 

> - Break large modules into smaller

> - Introduce the abstract module (contains only interfaces) to decouple the dependencies and improve the parallel building

- Leverage [Mainframer](https://github.com/buildfoundation/mainframer) to have the remote build setup. We got a very excellent result [This may help if you look for a quick win solution and haven't applied it to your project yet]
- We are in the process of migrating to use Bazel, and we are very close to the goal.

**Reply by manask88** [reply]
@roscrazy thanks for your answer. I had not heared about Mainframer, but sounds interesting . It does seem like there is an integration with AS. So curious about your experience with it. Does the developer experience changed much compared to when you use AS to compile the app? 

btw, we also tested https://github.com/joaquim-verges/ProjectorAndroidStudio but didn't find really big gains that would justify using it.

How big your Android team / build team?
In our case, we are about 40 engs in total, but I'm the only build engineer. Bazel sounds interesting, but for now we are looking to stay with Gradle given that it seems like we have more fundamental issues to solve  first(modularization)

**Reply by roscrazy** [reply]
Hi @manask88,
We mostly use the [command line](https://github.com/buildfoundation/mainframer) version. So, the engineer will use the AS as usual. But when they want to build the project, they have to switch to the command line tool and run the mainframer build command. There are two main benefits.

- 2x faster for the build time for our project
- The building process happens on a remote machine. It will not affect your IDE experience/or your local machine during the build execution.

Now, we have quite an advanced setup with K8s for auto-scale. But before that, we created a few EC2 instances and assigned each of them to a set of engines. You can start with that approach.
My project has 100+ android engineers, and we have about four platform engineers.
For Bazel migration, we build a tool for auto migration and soon will open-source the tool. Maybe you could try Bazel after that

**Reply by roscrazy** [reply]
Hi @manask88 , quick question. I would like to try the remote IDE, asking question [here](https://github.com/MobileNativeFoundation/discussions/discussions/57) 
Could you share a bit of result on how do you test the projector? how is your experience?
Thanks.

**Comment by tinder-inakivillar** [comment]
@manask88 we are testing a forked version of the KGP 1.4.31 with the fix. 
Let you know in the next days how is going

**Reply by tinder-inakivillar** [reply]
Everything ok at the moment, our changes were based in:
https://github.com/JetBrains/kotlin/compare/v1.4.31...tinder-inakivillar:apply_fix_kapt_remote_cache_problem
